ledger/src/shredder.rs

```diff
@@ -391,18 +391,35 @@ impl Shredder {
     }
 
     /// Combines all shreds to recreate the original buffer
-    pub fn deshred(shreds: &[Shred]) -> Result<Vec<u8>, Error> {
-        let index = shreds.first().ok_or(TooFewDataShards)?.index();
-        let aligned = shreds.iter().zip(index..).all(|(s, i)| s.index() == i);
-        let data_complete = {
-            let shred = shreds.last().unwrap();
-            shred.data_complete() || shred.last_in_slot()
-        };
-        if !data_complete || !aligned {
+    pub fn deshred<I, T: Borrow<Shred>>(shreds: I) -> Result<Vec<u8>, Error>
+    where
+        I: IntoIterator<Item = T>,
+    {
+        let (data, _, data_complete) = shreds.into_iter().try_fold(
+            <(Vec<u8>, Option<u32>, bool)>::default(),
+            |(mut data, prev, data_complete), shred| {
+                // No trailing shreds if we have already observed
+                // DATA_COMPLETE_SHRED or LAST_SHRED_IN_SLOT.
+                if data_complete {
+                    return Err(Error::InvalidDeshredSet);
+                }
+                let shred = shred.borrow();
+                // Shreds' indices should be consecutive.
+                let index = Some(shred.index());
+                if let Some(prev) = prev {
+                    if prev.checked_add(1) != index {
+                        return Err(Error::from(TooFewDataShards));
+                    }
+                }
+                data.extend_from_slice(shred.data()?);
```

Comment:
Had to refresh myself on the specifics of `extend_from_slice()`; the [docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice) say:
> Iterates over the slice other, clones each element, and then appends it to this Vec.

The mention of "iterate and clone" initially had me worried. However, for a `Vec<T>` where `T` is `Copy`, a regular memcpy will occur so we're good:
https://doc.rust-lang.org/src/alloc/vec/spec_extend.rs.html#54-63

Reply:
My note above aside, `extend_from_slice()` will potentially re-allocate several times. Thoughts on pre-allocating the `Vec` with enough bytes initially (num_shreds * max_data_paylod_per_shred) ?

I guess one shortcoming of that approach is that we'd have some "wasted" bytes allocated. But on second thought, I think that is true if we let `Vec` grow by doubling its' capacity as needed. And our worst case "waste" for pre-allocating might have a tighter bound than standard re-alloc strategy

Reply:
vector does exponential growth, so this will only do logarithmic many number of re-allocations:
https://github.com/rust-lang/rust/blob/b2728d542/library/alloc/src/raw_vec.rs#L649-L651

So I don't think it is a major issue here.
Nonetheless, I have some followup patches that instead of concatenating the data buffers into a contiguous vector, return a wrapper around the shreds iterator which implements the `Read` trait and use `bincode::deserialize_from` to read from that `Read` interface.